### ISSUE TITLE
Improve how to retrieve a particular ServiceProvider

### ISSUE THREAD

**Issue by keilw** [unitsofmeasurement/unit-api#124]
While retrieving a particular `ServiceProvider` (other than the **current** one) is being solved in the best way Java 8 allows by #123, there are ways to improve and optimize this retrieval in higher Java versions. As Java 8 is the most common version (~80% in some surveys) we don't want to force Java 9, 10, 11 or above into the current API, but would like to improve that in a future update.

- [x] Address #114 
- [x] Bring Coverage back to Green
- [ ] Further improvements

Needs #113
Needs #122
Needs #123
Needs #129

**Comment by desruisseaux** [comment]
Depends on #113, which needs to be resolved first.

**Comment by thodorisbais** [comment]
Maybe this now needs an additional `in-progress` label?

**Comment by keilw** [comment]
It has, @desruisseaux please also note the 2nd bullet point, because **code coverage** dropped significantly and the Sonar Quality Gate Failed for now because coverage dropped 15%!

**Comment by desruisseaux** [comment]
Raising coverage requires adding more tests. Before doing so, it may be worth to reformat at least `ServiceProviderTest` as we did for the main source code, so that when I edit that file, it does not appear as if I modified the whole file.

**Comment by keilw** [comment]
Please do whatever is necessary or should it be a big problem then maybe others could help with those tests.

**Comment by thodorisbais** [comment]
@keilw I received no feedback for the Associate one; probably because we've started escalating my Full Membership together with Harold (from the JCP Team). That one (Full Membership) is already signed both from Oracle's and my sides and we are now waiting for my employer's signature, too. 

We gave it a try with my boss last week, but for some reason, he didn't receive any confirmation mail after registering for an Oracle account. On top of that, he's on a 10 days vacation (starting from tomorrow), so, probably, I need to repeat the entire procedure (and explanation) to his delegate. The good news, for now, is that I suggested my employer registering to JCP as a Partner and they are already interested, so, hopefully, we'll have some news over the next few days (I know, it's a sequence of events/coincidences that are postponing the entire process, but given the circumstances, we couldn't/cannot do anything faster/better; however, I just asked Harold whether it's possible to for the Associate one first, until my request for the Full Membership gets signed from all the parties.

Indeed, I was unfortunate enough the other day to watch a "Meet the JCP EC Candidates BOF" video of CodeOne, without having the chance to vote. All in all, good to hear so, because an Adopt-a-JSR JCP Award is definitely my/our goal for next year :-) .

**Comment by keilw** [comment]
@thodorisbais Well it depends, if you join as Full Member, you would get to vote for different candidates than under Associate Membership. Either way, we had a recent JSR 385 Member @Daniel-Dos win the Adopt award (although when they nominated him, other JSRs were mentioned because he had not joined 385), @otaviojava won it once and @rajmahendra explicitly for adopting JSR 363, so there's some potential for future awards ;-)

**Comment by thodorisbais** [comment]
Indeed, I see. After having taken a look at the past awards, I am thinking about involving more people from our JUG and making it a more serious effort. Let's see how it goes. :-)

**Comment by keilw** [comment]
@desruisseaux Also noticed (because `ServiceProvider` is duplicated now) that the API JAR grew to 43kB because of this ticket. I hope it won't get much bigger? The tiny flag in the enum (which as such is usually much bigger than other elements) makes no size difference. I really think if the absolute/relative distinction whatever we call it should just be a boolean, too, so we can avoid further growth. If there will never be more than 2 options, that seems enough. 

As for the META-INF folder, I can't say, if it has a major impact, but https://stackoverflow.com/questions/27382259/why-do-i-need-meta-inf-maven-dir-in-my-jar suggests, it may not be needed at runtime, unless someone also utilized a custom classloader that uses this information to dynamically load dependencies. There are not many JARs leaving it out, but a flag in the plugin exists. Frameworks built with Gradle like Spring are missing that folder, although at more than 1MB per JAR (`spring-core`) it barely makes a difference ;-)

**Comment by desruisseaux** [comment]
Will see later this week if I can refactor the code in a more compact way. `ServiceProvider` is duplicated for now, but will not be duplicated anymore after we use Java 9 specific API for the class in the `jdk9` directory.

My problem with the tiny flag in enum is not a matter of size, but a matter of conceptual relevance. But this is another topic (#139).

For the `META-INF/maven` folder, I have never seen any software making use of it. This folder is very easy to exclude.

**Comment by keilw** [comment]
It is currently applied, since e.g. some JBoss JARs also do not seem to use it, there is no notable harm to leave it out. It gives us roughly 3 kB back, but optimizing either `ServiceProvider` or others is still a good thing to do. The target environment may not be Java ME Embedded now, but we should still aim for a footprint that is small enough to fit into a Raspberry Pi Zero or similar "Edge Device".

**Comment by keilw** [comment]
@desruisseaux,  @thodorisbais What would you say is open here? The coverage has been fixed some while ago. What about the further improvements?

**Comment by desruisseaux** [comment]
I would need to refresh my mind. Which improvements are remaining?

**Comment by keilw** [comment]
I thought this was about a more memory or otherwise efficient way to retrieve them under Java 9 or above. Is that done in https://github.com/unitsofmeasurement/unit-api/blob/master/src/main/jdk9/javax/measure/spi/ServiceProvider.java? If so, I think 3 could be checked off, too. The logging has already been improved in the JDK9 version.

**Comment by thodorisbais** [comment]
I missed this somehow. I was also wondering what "further improvements" (3rd checkbox) stands for. Let's see what Martin thinks about that one, then. Hopefully, this one will be closed too, soon.

**Comment by thodorisbais** [comment]
Do we still need to keep this open?

**Comment by keilw** [comment]
@desruisseaux Do you see any improvements that would be necessary at this point?

**Comment by desruisseaux** [comment]
I just had a look to the code. There is a few items, I let you judge if they should be considered important or not:

* The code checks for `javax.inject.Named` and `javax.annotation.Priority` annotations, but I don't know if there is corresponding JUnit tests (I did not checked yet).
* The `getPriority()` method is redundant with the `javax.annotation.Priority` annotation. Do we want to keep the `getPriority()` method? (I'm neutral on that).
* If the `javax.inject.Named` annotation is not present, the code uses the `toString()` return value as a name. Is it safe? (an implementation could have verbose `toString()`, especially since the current `ServiceProvider` interface does not define a contract for `toString()`.) Should we provide a `getName()` method instead (it would be redundant with the annotation), or alternatively consider that providers without `Named` annotation are unnamed?

**Comment by keilw** [comment]
Thanks for summarizing what could be done. Although JUnit tests cannot hurt and ServiceProvider may be among the lesser covered parts, the overall coverage is not bad above 80%. @thodorisbais Any thoughts on those?

**Comment by thodorisbais** [comment]
First of all, the title of this issue is a bit generic, but that's not a big deal, since in the description we have clear requests, except for the last one.

That is, "further improvements" can mean anything and nothing at the same time. The points of @desruisseaux might be enough, but there's a chance we'll miss again something since the request itself ("further improvements") is a vague "acceptance criterion" for this issue.

Having said that, I would suggest the following:
- remove the "further improvements" checkbox from this issue
- close this issue
- create a new issue of low priority (tagged neither as PR nor as 2.0) with the items Martin mentioned

**Comment by thodorisbais** [comment]
@keilw shall we?

**Comment by keilw** [comment]
I think this one is good to close.
